group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1702,65 +2012,218 @@ private CoordinatorResult<Void, CoordinatorRecord> classicGroupJoinToConsumerGro
+            );
+
+            if (!subscriptionMetadata.equals(group.subscriptionMetadata())) {
+                log.info("[GroupId {}] Computed new subscription metadata: {}.",
+                    groupId, subscriptionMetadata);
+                bumpGroupEpoch = true;
+                replayRecords.add(newGroupSubscriptionMetadataRecord(groupId, subscriptionMetadata));
+            }
+
+            if (bumpGroupEpoch) {
+                groupEpoch += 1;
+                records.add(newGroupEpochRecord(groupId, groupEpoch, SHARE));
+                log.info("[GroupId {}] Bumped group epoch to {}.", groupId, groupEpoch);
+            }
+
+            group.setMetadataRefreshDeadline(currentTimeMs + shareGroupMetadataRefreshIntervalMs, groupEpoch);
+        }
+
+        // 2. Update the target assignment if the group epoch is larger than the target assignment epoch.
+        final int targetAssignmentEpoch;
+        final Assignment targetAssignment;
+
+        if (groupEpoch > group.assignmentEpoch()) {
+            targetAssignment = updateTargetAssignment(
+                group,
+                groupEpoch,
+                updatedMember,
+                subscriptionMetadata,
+                subscriptionType,
+                replayRecords
+            );
+            targetAssignmentEpoch = groupEpoch;
+        } else {
+            targetAssignmentEpoch = group.assignmentEpoch();
+            targetAssignment = group.targetAssignment(updatedMember.memberId());
+        }
+
+        // 3. Reconcile the member's assignment with the target assignment if the member is not
+        // fully reconciled yet.
+        updatedMember = maybeReconcile(
+            groupId,
+            updatedMember,
+            targetAssignmentEpoch,
+            targetAssignment,
+            replayRecords
+        );
+
+        scheduleShareGroupSessionTimeout(groupId, memberId);
+
+        // Prepare the response.
+        ShareGroupHeartbeatResponseData response = new ShareGroupHeartbeatResponseData()
+            .setMemberId(updatedMember.memberId())
+            .setMemberEpoch(updatedMember.memberEpoch())
+            .setHeartbeatIntervalMs(shareGroupHeartbeatIntervalMs);
+
+        // The assignment is only provided in the following cases:
+        // 1. The member just joined or rejoined to group (epoch equals to zero);
+        // 2. The member's assignment has been updated.
+        if (memberEpoch == 0 || hasAssignedPartitionsChanged(member, updatedMember)) {
+            response.setAssignment(createShareGroupResponseAssignment(updatedMember));
+        }
+
+        /*
+         For a consumer group, below updates would occur as a result of applying (replaying) a member
+         or group persisted record. Because a share group doesn't persist additional records, hence we
+         need to replay the generated records here.
+        */
+        replayRecords.forEach(this::replay);
+        return new CoordinatorResult<>(records, response);
```

Comment:
I believe that this does not fully work as you expect. For instance, if you have records in `replayRecords` that you replay but without any records in `records`, the changes may be rolled back if the next write operation executed by the coordinator fails. This happens because the coordinator runtime does not know about those changes in the timeline data structures and it will revert to the last written offset which is computed based on the previous write operation. This could lead to tricky state inconsistencies. I am not sure if it matters in this particular case but it is for sure confusing.

Reply:
What you mean is that the whole replaying records should be atomic else soft updates might be overwritten in case of failures. Also you mentioned coordinator runtime should be made aware of any timeline changes hence does that mean soft updates should not be perfromed? Then probably we might have to think about persisting other records as well like Consumer Group.